src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngSymmetricAlgorithmCore.cs

```diff
@@ -0,0 +1,194 @@
+        {
+            _algorithm = algorithm;
+            _outer = outer;
+
+            _keyName = null; // Setting _keyName to null signifies that this object is based on a plaintext key, not a stored CNG key.
+            _provider = null;
+            _optionOptions = CngKeyOpenOptions.None;
+        }
+
+        /// <summary>
+        /// Constructs the core to use a stored CNG key. 
+        /// </summary>
+        public CngSymmetricAlgorithmCore(string algorithm, ICngSymmetricAlgorithm outer, string keyName, CngProvider provider, CngKeyOpenOptions openOptions)
+        {
+            if (keyName == null)
+                throw new ArgumentNullException("keyName");
+            if (provider == null)
+                throw new ArgumentNullException("provider");
+
+            _algorithm = algorithm;
+            _outer = outer;
+
+            _keyName = keyName;
+            _provider = provider;
+            _optionOptions = openOptions;
+
+            using (CngKey cngKey = ProduceCngKey())
+            {
+                CngAlgorithm actualAlgorithm = cngKey.Algorithm;
+                if (algorithm != actualAlgorithm.Algorithm)
+                    throw new CryptographicException(SR.Format(SR.Cryptography_CngKeyWrongAlgorithm, actualAlgorithm.Algorithm, algorithm));
+
+                _outer.BaseKeySize = cngKey.KeySize;
+            }
+        }
+
+        /// <summary>
+        /// Note! This can and likely will throw if the algorithm was given a hardware-based key.
+        /// </summary>
+        public byte[] GetKeyIfExportable()
+        {
+            if (KeyInPlainText)
+            {
+                return _outer.BaseKey;
+            }
+            else
+            {
+                using (CngKey cngKey = ProduceCngKey())
+                {
+                    return cngKey.GetSymmetricKeyDataIfExportable(_algorithm);
+                }
+            }
+        }
+
+        public void SetKey(byte[] key)
+        {
+            _outer.BaseKey = key;
+            _keyName = null; // Setting _keyName to null signifies that this object is now based on a plaintext key, not a stored CNG key.
+        }
+
+        public void SetKeySize(int keySize, ICngSymmetricAlgorithm outer)
+        {
+            // Warning: This gets invoked once before "this" is initialized, due to Aes(), DES(), etc., setting the KeySize property in their
+            // nullary constructor. That's why we require "outer" being passed as parameter.
+            Debug.Assert(_outer == null || _outer == outer);
+
+            outer.BaseKeySize = keySize;
+            _keyName = null; // Setting _keyName to null signifies that this object is now based on a plaintext key, not a stored CNG key.
+        }
+
+        public void GenerateKey()
+        {
+            byte[] key = new byte[_outer.BaseKeySize.BitSizeToByteSize()];
+            do
+            {
+                RandomNumberGenerator.Create().GetBytes(key);
```

Comment:
RandomNumberGenerator is IDisposable. We should Dispose of the instance when done with it, even if the current implementations have Dispose as a nop.